components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt

```diff
@@ -145,8 +147,10 @@ open class GleanInternalAPI internal constructor () {
         @Suppress("EXPERIMENTAL_API_USAGE")
         Dispatchers.API.flushQueuedInitialTasks()
 
-        // At this point, all metrics and events can be recorded.
-        ProcessLifecycleOwner.get().lifecycle.addObserver(gleanLifecycleObserver)
+        MainScope().launch {
```

Comment:
tbh I'm suprised and confused why Fenix is calling this on a different thread in the first place (since all other apps call it right from a main thread lifecycle callback of the `Application`). Is that really needed? Do you want to support that? Another route you could take is marking this with `@MainThread` or even use `ThreadUtils.assertOnUiThread()` to enforce it.

I fear that if you allow/support multiple threads that you may run into even more issues - specifically around [visibility](https://www.logicbig.com/tutorials/core-java-tutorial/java-multi-threading/visibility-volatile.html). One example is `initialized` in `Glean`: It's written in `initialize()` and not `synchronized` or `volatile`. If you set it to `true` from the thread that calls `initialize()` then a different thread calling `isInitialized()` may not see that value (yet) and still read `false`.

Reply:
I *think* Fenix is doing this for performance reasons but, honestly, I don't know :)

That's a good point with respect to the other implications. The `initialize` method might trigger I/O (off the main thread already!) so, thinking of this again, I'm not too sure this is strictly required. @mdboom we should probably talk to the Fenix people and check with them: maybe we can get along with enforcing init on the main thread.

Reply:
If there are performance reasons (are there?) then we'd want to "upstream" those optimizations into Glean, I assume. :)

Reply:
> 
> 
> If there are performance reasons (are there?) then we'd want to "upstream" those optimizations into Glean, I assume. :)

Yes, you're completely right. We'll check with them before moving on with this. Thanks for the feedback!